nixos/gnupg: add systemd configuration

corngood opened the pull request:
This depended on the systemd user configuration provided upstream in doc/examples.  However, this was all removed in:

https://github.com/gpg/gnupg/commit/eae28f1bd4a5632e8f8e85b7248d1c4d4a10a5ed

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

corngood committed:
nixos/gnupg: add systemd configuration

This depended on the systemd user configuration provided upstream in
doc/examples.  However, this was all removed in:

https://github.com/gpg/gnupg/commit/eae28f1bd4a5632e8f8e85b7248d1c4d4a10a5ed

rnhmjoj commented:
Yes, there is a race and I think the solution is for gnupg to just stop trying to manage the agent itself and let systemd (or any init system, for that matter) do its job. Unfortunately upstream went the other way.

This PR is a reasonable solution, however they officially dropped systemd and already hinted that `--pinentry-program` should not be used and may be removed. So, I would save ourselves some future troubles by configuring `pinentry-program` with /etc/gnupg.conf and let gnupg start the agent.

corngood commented:
That sounds reasonable.  I'll set it up that way instead...

On 10 May 2023 16:52:12 GMT-03:00, Michele Guerini Rocco ***@***.***> wrote:
>Yes, there is a race and I think the solution is for gnupg to just stop trying to manage the agent itself and let systemd (or any init system, for that matter) do its job. Unfortunately upstream went the other way.
>
>This PR is a reasonable solution, however they officially dropped systemd and already hinted that `--pinentry-program` should not be used and may be removed. So, I would save ourselves some future troubles by configuring `pinentry-program` with /etc/gnupg.conf and let gnupg start the agent.
>
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/231108#issuecomment-1542722975
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

corngood committed:
gnupg: fix tests

rnhmjoj commented on pkgs/tools/security/gnupg/24.nix:
```diff
@@ -85,7 +86,7 @@ stdenv.mkDerivation rec {
   enableParallelBuilding = true;
 
-  passthru.tests.connman = lib.nixosTests.gnupg;
+  passthru.tests.connman = nixosTests.gnupg;
```
I think this is the result of me copy-pasting and forgetting to change the name.
Can you fix the name?

rnhmjoj commented on nixos/modules/programs/gnupg.nix:
```diff
@@ -95,29 +95,101 @@ in
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/gpg-agent --supervised \
+            --pinentry-program ${pkgs.pinentry.${cfg.agent.pinentryFlavor}}/bin/pinentry
```
I would really avoid this now that we can: setting the pinentry program for the systemd service only has been a constant source of pain for a number of users. See all the issues we have about pinentry[1][2][3][4].
Upstream also said `--pinentry-program` is only meant for testing: you should set the path from the gnupg configuration file

If you want to preserve the systemd units, that ok, but at least make an option to enable them.

[1]: https://discourse.nixos.org/t/cant-make-gpg-agent-work-with-ssh-on-a-non-gui-system/24750
[2]: https://discourse.nixos.org/t/cant-get-gnupg-to-work-no-pinentry/15373
[3]: https://github.com/NixOS/nixpkgs/issues/35464
[4]: https://github.com/NixOS/nixpkgs/issues/210375

corngood replied:
I'm all for removing the systemd units.  My only worry is: will we be breaking anyone by removing the socket activation?  Like maybe ssh-agent compatibility?

It may still be a good idea regardless, but we'd probably need some documentation.

SuperSandro2000 replied:
We should probably add a changelog entry either way because this is most likely going to cause regressions and they are hard to debug with gnupg shit error messages ala doesn't work.

corngood replied:
I've moved this to /etc/

corngood committed:
gnupg: fix test attribute key

corngood commented:
I think my recommendation would be to get this change in while we consider the consequences of dropping the systemd units.  Then it'll be in the history in case people want to revert to it in the future, and we may even want to support both methods.

rnhmjoj reviewed (changes requested):
As I already said, I'm ok with keeping the unit files provided pinentry-program is set from `/etc/gnupg/gpg.conf`, not the command line because that is too brittle.

corngood commented:
@rnhmjoj sorry, I misunderstood.  I'll make that change.

corngood committed:
nixos/gpg-agent: move pinentry-program to /etc/gnupg/gpg-agent.conf

corngood commented:
Updated with:

- moved pinentry-program to /etc/gnupg/gpg-agent.conf
- removed the doc/examples stuff from gnupg

The later is needed for gnupg < 2.4.1, and it's been reverted to 2.4.0 in master.  Perhaps this should go in staging?

rnhmjoj commented:
Looks good. You can probably remove this line:
```
    systemd.packages = [ cfg.package ];
```
as we don't rely on (potential, at this point) upstream units anymore.

corngood commented:
@rnhmjoj Good catch, now we can leave:

https://github.com/NixOS/nixpkgs/pull/231108/commits/4bc7ddc9ca84a4e8c8fec9b07118b95de435e12f

until we upgrade gpg again, which will reduce the number of rebuilds here drastically.

Updated with that change.

rnhmjoj reviewed (approved):
I checked the definitions against the previously supplied units and it seems ok.
The test is passing, so let's not delay the update any longer and merge this now.

corngood commented:
I updated #231110 to include merging this into staging, and reverting the 2.4.1 rollback.

I'm not sure if that's the right approach to take, but if so, it's there.
